src/shogun/lib/parameter_observers/ParameterObserver.h

```diff
@@ -103,17 +134,32 @@ namespace shogun
 		virtual void on_complete() = 0;
 
 		/**
-		 * Method useful to empty the observer from
-		 * obseverd value it may have stored.
+		 * Get the name of this class
+		 * @return name as a string
 		 */
-		virtual void clear(){};
+		virtual const char* get_name() const
+		{
+			return "ParameterObserver";
+		}
 
 	protected:
+		/**
+		 * Implementation of the on_next method which will be needed
+		 * in order to process the observed value
+		 * @param value the observed value
+		 */
+		virtual void on_next_impl(const TimedObservedValue& value) = 0;
+
 		/**
 		 * List of parameter's names we want to monitor
 		 */
 		std::vector<std::string> m_parameters;
 
+		/**
+		 * Observations recorded each time we compute on_next()
+		 */
+		std::vector<Any> m_observations;
```

Comment:
All the observer will store the values they capture. For now, they will be kept as `Any` values since they may be really different from each other (a scalar, a vector, etc.). Does it seem clean enough?

Reply:
Not sure. We know that all observed values have the same type or? So could template this? Not sure...

Reply:
Okay, so I've come up with another idea which I think it's the most viable option (and also it does not expose `Any` to the final users). Basically, the observers will have an `std::vector<ObservedValue>` parameter which will contain all the observation captured so far (we will not use the `Any`). If I want to access it from the interface I can do:
```c++
ParameterObserver pm = parameter_observer("ParameterObserverCV")
ObservedValue obs = pm.get_observations(1)
```
`ObservedValue` will contain both the observed `Any` value and the name of it as registered parameters. Therefore, we could extract the value itself from the `ObservedValue` like:
```c++
obs.get_real("value") // if we know that the value is real
obs.get_real("value") // if it is a string...
```
The only possible issue is that inside `ObservedValue` we would need to register (with `SG_ADD`) an `Any` value, since we cannot know a-priori which will be the type of the observed value.

Reply:
In order to expose the observations (aka `std::vector<ObservedValue>`) we can:
  * Use the current method (`get_observations(int i)`) which won't need `std::vector<ObservedValue>` to be registered with `SG_ADD`;
  * Use `get` and add a new SWIG getter like `%template(get_observation) CSGObject::get<std::vector<ObservedValue>, void>;`. We would need also to add a new type called `ObservedVector` which will correspond to `std::vector<ObservedValue>`.

Reply:
I think in our discussion, we discussed the first option you described. I think it might work!